scene/resources/height_map_shape.cpp

```diff
@@ -0,0 +1,196 @@
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "height_map_shape.h"
+#include "servers/physics_server.h"
+
+Vector<Vector3> HeightMapShape::_gen_debug_mesh_lines() {
+	Vector<Vector3> points;
+
+	// This will be slow for large maps...
+	// also we'll have to figure out how well bullet centers this shape...
+
+	Vector2 size(map_width - 1, map_depth - 1);
+	Vector2 start = size * -0.5;
+	int offset = 0;
+
+	PoolRealArray::Read r = map_data.read();
+
+	for (int d = 0; d < map_depth; d++) {
+		Vector3 height(start.x, 0.0, start.y);
+
+		for (int w = 0; w < map_width; w++) {
+			height.y = r[offset++];
+
+			if (w != map_width - 1) {
+				points.push_back(height);
+				points.push_back(Vector3(height.x + 1.0, r[offset], height.z));
+			}
+
+			if (d != map_depth - 1) {
+				points.push_back(height);
+				points.push_back(Vector3(height.x, r[offset + map_width - 1], height.z + 1.0));
+			}
+
+			height.x += 1.0;
+		}
+
+		start.y += 1.0;
+	}
+
+	return points;
+}
+
+void HeightMapShape::_update_shape() {
+
+	Dictionary d;
+	d["width"] = map_width;
+	d["depth"] = map_depth;
+	d["heights"] = map_data;
+	d["min_height"] = min_height;
+	d["max_height"] = max_height;
+	PhysicsServer::get_singleton()->shape_set_data(get_shape(), d);
+}
+
+void HeightMapShape::set_map_width(int p_new) {
+	if (p_new < 1) {
+		// ignore
+	} else if (map_width != p_new) {
+		int was_size = map_width * map_depth;
+		map_width = p_new;
+
+		int new_size = map_width * map_depth;
+		map_data.resize(map_width * map_depth);
+
+		PoolRealArray::Write w = map_data.write();
+		while (was_size < new_size) {
+			w[was_size++] = 0.0;
+		}
+
+		_update_shape();
+		notify_change_to_owners();
+		_change_notify("map_width");
+		_change_notify("map_data");
+	}
+}
+
+int HeightMapShape::get_map_width() const {
+	return map_width;
+}
+
+void HeightMapShape::set_map_depth(int p_new) {
+	if (p_new < 1) {
+		// ignore
+	} else if (map_depth != p_new) {
+		int was_size = map_width * map_depth;
+		map_depth = p_new;
+
+		int new_size = map_width * map_depth;
+		map_data.resize(new_size);
+
+		PoolRealArray::Write w = map_data.write();
+		while (was_size < new_size) {
+			w[was_size++] = 0.0;
+		}
+
+		_update_shape();
+		notify_change_to_owners();
+		_change_notify("map_depth");
+		_change_notify("map_data");
+	}
+}
+
+int HeightMapShape::get_map_depth() const {
+	return map_depth;
+}
+
+void HeightMapShape::set_map_data(PoolRealArray p_new) {
+	int size = (map_width * map_depth);
+	if (p_new.size() != size) {
+		// fail
+		return;
+	}
+
+	// copy
+	PoolRealArray::Write w = map_data.write();
+	PoolRealArray::Read r = p_new.read();
+	for (int i = 0; i < size; i++) {
+		float val = r[i];
+		w[i] = val;
+		if (i == 0) {
+			min_height = val;
+			max_height = val;
```

Comment:
min and max are calculated too in PhysicsServer if not provided, although I wonder if it would make sense to expose a getter to obtain them in the shape API